### PR TITLE
perf: replace samples map with a slice

### PR DESCRIPTION
**Pull request by kruskall**
## Motivation/summary

Reduce allocations by replacing the samples map with a slice
with perfect size.

Benchmark on heavy.ndjson:

```
name                       old time/op    new time/op    delta
ReadBatch/heavy.ndjson-16    11.2ms ± 9%    10.5ms ± 8%   -5.85%  (p=0.029 n=10+10)

name                       old speed      new speed      delta
ReadBatch/heavy.ndjson-16  35.7MB/s ± 9%  37.9MB/s ± 8%   +6.20%  (p=0.029 n=10+10)

name                       old alloc/op   new alloc/op   delta
ReadBatch/heavy.ndjson-16    1.28MB ± 0%    0.83MB ± 0%  -35.09%  (p=0.000 n=10+10)

name                       old allocs/op  new allocs/op  delta
ReadBatch/heavy.ndjson-16     23.1k ± 0%     22.6k ± 0%   -2.09%  (p=0.000 n=10+9)
```

## Checklist

- [ ] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/main/CHANGELOG.asciidoc)
- [ ] Update [package changelog.yml](https://github.com/elastic/apm-server/blob/main/apmpackage/apm/changelog.yml) (only if changes to `apmpackage` have been made)
- [ ] Documentation has been updated

For functional changes, consider:
- Is it observable through the addition of either **logging** or **metrics**?
- Is its use being published in **telemetry** to enable product improvement?
- Have system tests been added to avoid regression?

### CODE REVIEW

File: internal/model/metricset.go [hunk 1]
```diff
@@ -70,6 +70,9 @@ type MetricsetSample struct {
 	// If Type is unspecified or invalid, it will be ignored.
 	Type MetricType
 
+	// Name holds the name of the metric type.
```
Comment (axw) [inline internal/model/metricset.go:73]:
```suggestion
	// Name holds the metric name.
	//
	// Name is required.
```

**Review by axw — COMMENTED** [review]
Looks great! There's a couple of unrelated test failures (I'm looking into the service metrics one now), but there's one that seems likely to be related: `processor/otel.TestConsumeMetrics_JVM`

**Comment by kruskall** [comment]
Flaky tests should be fixed 🏓

**Pull request merged**